bd-crash-handler/src/uploader.rs

```diff
@@ -0,0 +1,282 @@
+            return Ok(());
+          }
+          Some((uuid, contents)) = self.upload_queued_rx.recv() => {
+            log::debug!("tracking artifact: {uuid} for upload");
+            self.track_new_upload(uuid, contents).await?;          }
+
+        }
+        continue;
+      };
+
+      let file_path = REPORT_DIRECTORY.join(&next.name);
+
+      if next.pending_intent_negotation {
+        let Ok(decision) = self.perform_intent_negotiation(&next.name).await else {
+          log::debug!("intent negotiation failed due to data channel closing, exiting task");
+          return Ok(());
+        };
+
+        match decision {
+          IntentDecision::Drop => {
+            self.file_system.delete_file(&file_path).await?;
+            self.write_index().await?;
+          },
+          IntentDecision::UploadImmediately => {
+            log::debug!("uploading artifact: {}", next.name);
+            // Mark the file as being ready for uploads and persist this to the index.
+            next.pending_intent_negotation = false;
+            self.index.push_front(next);
+            self.write_index().await?;
+          },
+        }
+
+        continue;
+      }
+
+      let Ok(contents) = self.file_system.read_file(&file_path).await else {
+        log::debug!(
+          "failed to read file for artifact {}, deleting and removing from index",
+          next.name
+        );
+        self.file_system.delete_file(&file_path).await?;
+        self.write_index().await?;
+
+        return Ok(());
+      };
+
+
+      self.upload_artifact(contents, &next.name).await?;
+
+      self.file_system.delete_file(&file_path).await?;
+      self.write_index().await?;
+    }
+  }
+
+  // Initialize the uploader from the index file on disk.
+  async fn initialize(&mut self) -> anyhow::Result<()> {
+    let path = REPORT_DIRECTORY.join(&*REPORT_INDEX_FILE);
+    log::debug!("initializing index: {}", path.display());
+    self.index = match self
+      .file_system
+      .read_file(&path)
+      .await
+      .and_then(|contents| read_compressed_protobuf::<ArtifactUploadIndex>(&contents))
+    {
+      Ok(index) => index,
+      Err(e) => {
+        log::debug!("unable to open index: {e}");
+        log::debug!("creating new index");
+
+        self.file_system.remove_dir(&REPORT_DIRECTORY).await?;
+        self.file_system.create_dir(&REPORT_DIRECTORY).await?;
+        ArtifactUploadIndex::default()
+      },
+    }
+    .artifact
+    .into_iter()
+    .collect();
+
+    Ok(())
+  }
+
+  async fn track_new_upload(&mut self, uuid: Uuid, contents: Vec<u8>) -> anyhow::Result<()> {
+    let uuid = uuid.to_string();
+    self.index.push_back(Artifact {
+      name: uuid.clone(),
+      time: self.time_provider.now().into_proto(),
+      pending_intent_negotation: true,
+      ..Default::default()
+    });
+
+    self
+      .file_system
+      .write_file(&REPORT_DIRECTORY.join(uuid), &contents)
+      .await?;
```

Comment:
I think at this point every file we write has to have at least CRC protection. We don't necessarily have to use the compression approach (that is probably overkill and perhaps we should redo some of that to just write the file + a CRC at the end), but I think we have to make sure that it is in place. So either this code has to do it or we have to assume the calling code is going to do it. Then we have to decide whether it will be checked on read back or we will assume the server is going to check it. We should discuss this offline with the team to figure out what we want to do here.